### PR TITLE
fix(kimi): consolidated kimi-native hardening — TUI interaction, forwarder lifecycle, usage & cost

### LINKED ISSUE #4437: kimi-native: sessions silently lose dispatched messages, strand mid-turn, and misreport runner disconnects — repro + audit
## Summary

`kimi-native` harness sub-agent sessions are markedly flakier than `claude-native`: dispatched messages are silently lost, sessions strand mid-turn or go `idle` without doing anything, and routine runner-tunnel blips get stamped `runner_disconnected`. A live reproduction plus a source audit traced this to a set of concrete defects in the kimi bridge/forwarder and two harness-agnostic server/runner weaknesses. Related launch-flag gaps were filed separately (#4394, #4395); this issue covers everything else.

## Live reproduction (captured 2026-08-08, kimi CLI 0.34.0, macOS runner)

Two `kimi-native` sub-agent sessions were dispatched simultaneously by an orchestrator. Both stranded identically:

1. `09:01:30 / 09:01:34` — kimi terminal auto-created, wire-log forwarder started (runner log).
2. `09:01:32 / 09:01:35` — turn forwarded to the harness; `inject_user_message` begins.
3. `~09:02:02 / ~09:02:05` — the bridge's readiness gate `_settle_pane` hits its **30s deadline** (`_TMUX_READY_TIMEOUT_S = 30.0`, `omnigent/kimi_native_bridge.py:37`) and **falls through silently** (`kimi_native_bridge.py:276-296`), blind-pasting into a still-booting TUI that discards the input.
4. `09:02:07 / 09:02:10` — kimi 0.34.0 finishes mounting its TUI (input box + `context:` footer) **~5s after the gate gave up** (proven by `kimi-code-home/workspaces.json` mtimes vs. runner log timestamps). First boot took **~36s**, including a `managed:kimi-code` OAuth refresh attempt.
5. The 5s paste-needle wait never sees the draft → silently skipped; the blind Enter lands in a now-mounted but **empty** input box (`kimi_native_bridge.py:358-366`).
6. `inject_user_message` reports success; `run_turn` yields an unconditional `TurnComplete` (`omnigent/inner/kimi_native_executor.py:91-96`); the session is marked **idle**.
7. End state observed: both TUIs at "No session yet — one will be created on your first message", `context: 0% (0/1M)`, empty input box; omnigent sessions `idle` with only a `resource_event` in history; the parent orchestrator waits forever; the forwarder polls forever for a `wire.jsonl` that will never exist.

No error surfaced anywhere at any layer.

### A. Message injection is unverified and its readiness gate fails silent (the repro above)
- `_settle_pane` has a single readiness marker (footer substring `"context:"`, `kimi_native_bridge.py:55`) and on timeout falls through and pastes anyway (`:276-296`); claude's equivalent (`_wait_for_claude_prompt_ready`) raises.
- The submit is one blind Enter after a ≤5s paste-needle wait (`kimi_native_bridge.py:354-366`); claude verifies the draft left the input box, retries Enter, and raises on failure (`omnigent/claude_native_bridge.py:2938-2971`).
- `run_turn` yields `TurnComplete` unconditionally after the paste (`inner/kimi_native_executor.py:91-96`), so a swallowed message looks like a successful empty turn.

### B. Turn completion has exactly one source and no failure/interrupt edge
- Idle is derived solely from a `step.end` wire-log row with `finishReason == "end_turn"` (`omnigent/kimi_native_forwarder.py:216-229`). No failed edge, no interrupt edge (Escape via `runner/native/interrupt.py:179` writes no `end_turn`), no PTY-quiescence fallback (claude has all three, `claude_native_forwarder.py:253-265, 1714-1730`), no pane-death edge.
- Any kimi crash/exit mid-turn, interrupt, or non-`end_turn` finish reason strands the session in `running` forever.

### C. Forwarder auth never refreshes — the session's only idle signal dies on token lapse
- `_auto_create_kimi_terminal` snapshots a one-shot bearer (`runner/native/orchestration.py:3580-3587`); `supervise_kimi_forwarder` doesn't accept an `auth` parameter (`kimi_native_forwarder.py:451-461`), unlike qwen (`orchestration.py:3458-3477`) and claude.
- On 401 the poll loop breaks **without advancing `last_line`** (`kimi_native_forwarder.py:443-447`) and retries the same item every 250ms forever; it never calls `record_post_failure` (`omnigent/_native_forwarder_health.py`), so the failure is invisible.

### D. Wire-log discovery is a recency heuristic over a globally shared store
- Every session home symlinks the user's **global** kimi `sessions/` store (`omnigent/kimi_native_credentials.py:120-131`); discovery picks the newest-mtime `wire.jsonl` ≥ launch−10s, with `workDir` filtering only after `session_index.jsonl` lists the session — which it doesn't until the first turn (`kimi_native_forwarder.py:148-179`).
- Two kimi sessions launched near-simultaneously (the fan-out case) can adopt each other's wire logs: crossed transcripts, `turn_end` credited to the wrong session, or a forwarder that never finds its log.

### E. No resume; the pane reaper silently amputates history
- Kimi records no resumable chat id (`omnigent/kimi_native.py:95-107, 366-373`); the runner self-heal doc promises `--resume` semantics that kimi doesn't have (`runner/app.py:6871-6882`). The 1-hour idle pane reaper includes kimi (`omnigent/terminals/pane_reaper.py:51-69`), so an idle-1h conversation transparently restarts context-free.

### F. Approval plumbing: hook long-poll exceeds kimi's hook timeout; keystroke scrape is wording-pinned
- The permission hook long-polls the web verdict for 3600s (`omnigent/kimi_native_hook.py:73`) but kimi kills hooks at 600s (`kimi_native_credentials.py:76-92`) — approvals clicked after 10 minutes are lost and the pane stalls.
- Web Approve/Deny types `1`/`3` only if the literal `"Approve once"` is on screen (`kimi_native_bridge.py:384-427`); on mismatch it silently no-ops (`kimi_native_hook.py:243-251`).

### G. Server/runner: disconnect attribution and backpressure (harness-agnostic, amplifies A–D)
- The disconnect reconciler fails **mid-turn** sessions with `runner_disconnected` on tunnel loss (`omnigent/server/app.py:2069-2132`) — so every defect above that falsely leaves a session mid-turn converts routine tunnel reconnects into "kimi disconnects".
- `on_runner_disconnect` can run twice per teardown (inner `finally` at `server/routes/runner_tunnel.py:561-571` plus outer `except` layers at `:580-601`); needs one idempotent, generation-aware path.
- The session-SSE relay stamps any `httpx.HTTPError`/`ConnectionError` as `runner_disconnected` (`server/routes/_sessions/orchestration.py:5873-5899`), misattributing session-stream failures to the runner.
- Terminal output pumping uses unbounded asyncio queues with synchronous parsing between awaits (`omnigent/terminals/control_bridge.py:474-509, 528-546`, `omnigent/terminals/ws_bridge.py:592-613`, `runner/transports/ws_tunnel/registry.py:275`); a TUI output flood can starve the runner loop → missed keepalives or process death → a real disconnect.

### H. Minor parity notes
- Forwarder re-reads the whole wire log every 250ms (`kimi_native_forwarder.py:266-290`) vs. claude's byte offsets.
- Kimi launch-arg derivation for server-spawned sub-agents (#4394) is fixed on a separate branch; not duplicated here.

## Fix plan (PRs)

- [ ] **PR 1 — kimi TUI interaction hardening** — #4439 (findings A, F): readiness gate raises instead of falling through, longer boot budget, ≥2 readiness markers; verified submit with retry; executor propagates inject failure as a failed turn; approval long-poll capped below kimi's hook timeout; approval scrape matches multiple markers and logs loudly on miss.
- [ ] **PR 2 — kimi forwarder & session lifecycle** — #4441 (findings B, C, D, E, H): refresh-capable forwarder auth + forwarder-health attribution; failure/interrupt/pane-death turn edges + quiescence fallback; private per-session sessions dir; byte-offset wire reads; pane-reaper exemption or resume story.
- [ ] **PR 3 — server disconnect attribution** — #4438 (finding G, first three bullets): idempotent generation-aware teardown; stop stamping session-stream errors as `runner_disconnected`.
- [ ] **PR 4 — terminal output backpressure** — #4440 (finding G, last bullet): bounded queues, drop/coalesce policy, yield after bounded parse batches.

## Environment

- Omnigent @ `7ab46cf47` (branch `fix/kimi-flakiness`), macOS runner, kimi CLI 0.34.0 (`~/.kimi-code/bin/kimi`), tmux-terminal native launches.

### LINKED ISSUE #4479: [Bug] kimi/kimi-native sessions never report model, token usage, or cost — cost badge empty and cost-budget policies hard-DENY
## Problem

Kimi sessions (`kimi` headless and `kimi-native` interactive harnesses) surface **no model name, token usage, or cost** anywhere in omnigent — no cost badge, no per-model attribution, no context ring. Worse than cosmetic: the `cost_budget` / `subagent_cost_budget` policies hard-DENY tool calls when `total_cost_usd` was never populated (`omnigent/policies/builtins/cost.py:118-132`), so kimi sessions break budget enforcement, not just display. claude-native and codex-native both report all three.

## Root cause (Kimi Code 0.34.0)

* **Headless** `kimi` launches `kimi --output-format stream-json -p …` and parses stdout JSONL (`omnigent/inner/kimi_executor.py:460-508`). Kimi's stream-json protocol emits *only* assistant/tool/meta records — usage, model, and cost fields do not exist in that stream (upstream `apps/kimi-code/src/cli/prompt-render.ts:77-99,262-265`). `TurnComplete` is emitted without `usage` (`kimi_executor.py:554-556`) even though the runtime accepts it (`omnigent/inner/executor.py:151-177`).
* `kimi-native` forwarder already tails Kimi's persisted `wire.jsonl` but maps only prompt/content/step-end records (`omnigent/kimi_native_forwarder.py:195-263`) — it reads the file that contains the data and drops the records that carry it.
* The wire log **has everything needed**:
  * `usage.record` — `{model, usage: {inputOther, output, inputCacheRead, inputCacheCreation}, usageScope: "turn"}`
  * `llm.request` — `{model, modelAlias, maxTokens, provider, …}` (provider-resolved id, e.g. `system.ai.kimi-k3`)
* Kimi emits **no monetary data** in any mode — same as codex, which sends tokens-only and lets the server price them via the model catalog.

## Proposed fix (grounded in the claude-native / codex-native pattern)

No server changes. Both kimi paths adopt the existing generic contract (`POST /v1/sessions/{id}/events`):

1. **kimi-native forwarder**: map `usage.record` → coalesced `external_session_usage` posts with cumulative token fields (`cumulative_input_tokens` = Σ inputOther+inputCacheCreation, `cumulative_cache_read_input_tokens`, `cumulative_output_tokens`) and `model` riding along on every post (codex `_SessionUsageCoalescer` pattern); `context_tokens` from the latest record's occupancy; `context_window` from omnigent's model catalog (never `llm.request.maxTokens` — that's max *output* tokens). Map `llm.request` → deduped `external_model_change`, not seeded at spawn (mirrors codex `_sync_model_change` so the real spawn model lands in `model_override`).
2. **Headless** `kimi` **executor**: locate the session's `wire.jsonl` via the `session.resume_hint` session id, keep a per-session byte-offset checkpoint (resume-safe), and sum the turn's `usage.record`s into `TurnComplete.usage`.
3. **Error handling**: strictly best-effort — missing/malformed wire log or schema drift logs once and emits nothing; never fails a turn. 0.34.0 wire schema pinned in test fixtures.

Full design doc: `docs/superpowers/specs/2026-08-09-kimi-usage-cost-design.md` (will land with the PR).

## Follow-ups (out of scope)

* Upstream a structured usage record to MoonshotAI/kimi-cli stream-json; then the headless wire.jsonl read can be deleted.
* Kimi/Moonshot pricing entries in the model catalog if absent (cf. omnigent-ai/omnigent#3144) — without them tokens persist but stay unpriced.

## Prior art

Searched issues/PRs for kimi usage/cost — none address this. Nearest non-overlapping: omnigent-ai/omnigent#1877/omnigent-ai/omnigent#2323 (tool-call cards), omnigent-ai/omnigent#4441 (forwarder hardening), omnigent-ai/omnigent#3144 (pricing ids).

### PR DESCRIPTION
**Pull request by btli**
## Related issue

Closes omnigent-ai/omnigent#4437. Also fixes #4479 (kimi model/usage/cost reporting), absorbed from #4481; the forwarder-lifecycle findings B, C, D, E, H from #4437 are absorbed from #4441.

## Summary

Kimi-native sessions were the flakiest lane: turns injected into an unready TUI, sessions stranded `running` forever after a kimi crash, parallel sessions adopting each other's transcripts, and no model / token / cost data at all (which made cost-budget policies hard-DENY every kimi tool call). This PR is the consolidated kimi hardening + accounting omnibus, in three layers:

**TUI interaction hardening (originally this PR)**

- Separate 120 s readiness budget waiting for kimi's input row + context footer (cold first boot takes ~36 s; the old 30 s gate pasted into an unready pane and reported success), with a typed error when the TUI never mounts.
- Verify pasted drafts appear and then leave the input box, retrying Enter before failing the turn with an actionable error.
- Bound approval polling below kimi's 600 s hook ceiling, re-park with the same elicitation id, and make approval-marker misses loud and typed; empty pane captures are retried before the menu is declared gone.

**Forwarder lifecycle, auth, and discovery (absorbed from #4441)**

- **Turn edges:** the top-level `turn.ended` wire record is the authoritative terminal signal (`completed`/`cancelled` → idle, `failed` → failed with the provider error surfaced; `step.end` finish reasons are never edges — failed/cancelled turns write no `step.end` at all, pinned by real-session fixtures under `tests/fixtures/kimi_wire/`). A dead kimi tmux pane mid-turn posts a failed edge; a 5-minute quiet-wire quiescence fallback closes edge-less turns, suppressed while a tool call is in flight (with a 30-minute hung-tool ceiling). Turn lifecycle is persisted with the cursor so fallbacks stay armed across forwarder restarts, and edges are deduped by id across restarts.
- **Auth + delivery:** refresh-capable `httpx.Auth` (mirroring qwen/claude) instead of a one-shot bearer snapshot; transient 4xx (401/403/408/409/425/429) retry with backoff, permanent 4xx poison-drop after 3 attempts so one bad line can't stall the tail forever; a dropped edge's status is remembered so the fallback close never softens a failure to idle; sustained edge-delivery failure alerts via rate-limited error logs.
- **Discovery:** each session home gets a private `sessions/` store + `session_index.jsonl` (no more symlinks to the global kimi store), so parallel kimi sessions can't adopt each other's wire logs; strict launch-epoch discovery.
- **Reads + reaper:** byte-offset incremental tailing (truncation rewind, partial trailing line deferred) instead of whole-file re-reads at 4 Hz; kimi exempted from the 1-hour pane idle reaper (no resumable chat id → a reaped pane is context-free).

**Model, token usage, and server-priced cost (absorbed from #4481)**

- The forwarder maps `usage.record` rows into cumulative, coalesced `external_session_usage` posts (input inclusive of cache reads per the server contract; cache-creation folds into input; `context_tokens` from the latest record; `context_window` resolved per model and omitted when unresolvable) and mirrors `llm.request` as a deduped `external_model_change` — so the cost badge, context ring, and cost-budget gate work on kimi sessions. Cumulative totals persist separately from the disposable wire cursor and survive terminal recreation; wire-cursor advancement is never gated on usage durability (recorded design ruling), and a per-poll sync retries failed usage posts even when no new wire rows arrive.
- The headless `kimi -p` executor sums each turn's `usage.record` rows into `TurnComplete.usage` with a byte-offset checkpoint and time-gated first read so resumes never re-bill history; the effective model is stamped into `usage["model"]`.
- `find_model_context_window()`: the existing window lookup minus the 128 K default, for callers that must omit rather than guess.
- **Kimi catalog pricing (consolidation addition):** the shared MLflow catalog already carries the kimi families (`moonshot.json`: kimi-k3 / kimi-k2.7-code / kimi-k2.6; `databricks.json`: databricks-kimi-k3), so `fetch_model_pricing()` now reaches them by routing: `kimi-*` ids resolve against the Moonshot catalog file, and Databricks Unity Catalog `system.ai.<model>` ids (what the kimi-native forwarder reports) map to their `databricks-<model>` serving row. The same routing resolves the context window for those ids. No pinned rate table.

ELI5: don't hand kimi a letter until its mailbox exists, and confirm the letter left the box. Kimi keeps a diary (`wire.jsonl`) of what it said, when each turn really ended, and every model call's token counts; we now tail that diary with a bookmark — recognizing crashes and interruptions as endings, giving each session its own diary shelf, renewing our library card — and tell the server, which prices the tokens (at kimi's posted rates when the catalog doesn't know the model) into the cost badge and budget gate.

```
web turn ► readiness: input row + footer ► paste ► draft visible ► Enter/retry ► draft gone
permission hook ► web poll ► empty window ► same elicitation id ► re-park ► approve/deny key

kimi TUI (tmux pane) ──appends──► wire.jsonl (session-private)
                                     │ byte-offset tail (250 ms)
                                     ▼
                              kimi forwarder ── refresh-capable auth ──► POST /v1/sessions/{id}/events
                               ├─ turn.ended completed/cancelled ─► idle       ├─ external_conversation_item
                               ├─ turn.ended failed ──────────────► failed     ├─ external_session_usage (cumulative)
                               ├─ pane died mid-turn ─────────────► failed     ├─ external_model_change (deduped)
                               ├─ wire quiet 5 min (no tool) ─────► idle       └─ external_session_status
                               └─ tool in flight quiet 30 min ────► failed
                                  cost: client-priced sum over per-model segments (cumulative_cost_usd);
                                  server token-prices at the current model as fallback (shared catalog)
kimi -p (headless) ─► wire.jsonl ─► executor ─► TurnComplete.usage {tokens, model}
```

## Design rulings

Recorded human rulings from the #4481 review tribunal — engines must not re-raise these:

- **`AP_CONTEXT_WINDOW_OVERRIDE` is honored by the forwarder's context-window lookup.** An explicit operator action documented as override-everything; uncatalogued kimi models are exactly its use case.
- **Wire-cursor advancement is never gated on usage-state durability (transcript liveness > usage durability).** A failed usage-state persist logs once, keeps totals in memory, and the per-poll sync retries delivery and the write; clamp-safe under the server's monotonic clamp.

## Test Plan

- `uv run pytest tests/test_kimi_native_forwarder.py tests/inner/test_kimi_harness.py tests/inner/test_kimi_native_executor.py tests/llms/test_context_window.py tests/runner/test_app_sessions_native_workflow_init.py tests/terminals/test_pane_reaper.py tests/test_kimi_native_credentials.py tests/test_kimi_native_hook.py` — **459 passed** (the union of all three source PRs' suites plus the review rounds' new tests).
- `pre-commit run --files <all changed files>` — all hooks pass except pyrefly, whose only errors are pre-existing environmental missing-imports (optional opentelemetry extras) in untouched `omnigent/runtime/telemetry.py`.
- **Kimi catalog routing verification:** unit tests cover the forwarder-reported ids (`system.ai.kimi-k3` → databricks row, `kimi-k3-databricks` → Moonshot family row, `system.ai.kimi-k2.7-code` → Moonshot exact row) against a fixture catalog, mutation-checked red without the routing change; also verified against the live MLflow catalog (K3 3.00/15.00/0.30, K2.7 Code 0.95/4.00/0.19, K2.6 0.95/4.00/0.16 per MTok, 1M window for the Databricks row).
- Merge reconciliation covered by tests: the real-wire fixtures now also exercise the usage/model parser (`failed.jsonl` yields `message → model → turn_failed`); the usage-sync live-loop tests run against the byte-offset loop from #4441.
- Adversarial review round 2 (Grok/Kimi/Codex, 7 findings, several probed empirically) applied: per-segment ACCRUED cost (rate changes price only new tokens — no clamp-freeze or retroactive overcharge), per-wire billed watermarks (A→B→A discovery bounce never re-bills), historical `turn.ended` edges gated on the launch-epoch floor, per-turn fallback edge ids (no strand under a stalled delivery cursor), first-read billing floor for empty/unreadable wires, torn-trailing-line-safe headless checkpointing, and request→usage-paired model attribution — each with mutation-checked tests.
- Adversarial review round 1 (3 reviewers, 6 findings) applied: per-model segment-priced `cumulative_cost_usd` (mid-session model switches no longer mis-charge under the server's monotonic clamp), generation-dedupe reset on same-path wire recreation, first-read `llm.request` attribution without a fabricated `time`, persisted in-flight assistant text, wire-independent usage retry + fallback-close sync, and the pending-first-read billing floor — each with mutation-checked tests.
- Deflaked `test_empty_capture_is_retried_before_menu_disappears` (patches the settle timeout/poll interval like the rest of its class; the real 0.5 s window left ~0.2 s of scheduling margin under load).

## Demo

- [Demo video (44s, MP4)](https://github.com/btli/omnigent/releases/download/pr-demos-2026-08-29/demo-4439-kimi.mp4) — before/after on real kimi wire fixtures: **before** (`main`) a failed kimi turn stays `RUNNING` forever with no model/token/cost data; **after** (this branch) the 401 surfaces as a failed edge, the model is mirrored, token/cost usage posts ($0.2928 at posted K3 rates), and a dead pane's newline-less completed tail closes as idle instead of stranding.
- Validating checks: `Pytest (inner-rest)` / `Pytest (runtime-harnesses)` cover the forwarder/executor suites (459 tests, see Test Plan); the real-wire fixtures under `tests/fixtures/kimi_wire/` pin the turn-edge and usage semantics shown in the clip.

## Type of change

- [x] Bug fix
- [x] Feature
- [ ] UI / frontend change
- [ ] Refactor / chore
- [ ] Docs
- [ ] Test / CI
- [ ] Breaking change

## Test coverage

- [x] Unit tests added / updated
- [x] Integration tests added / updated
- [ ] E2E tests added / updated
- [ ] Manual verification completed
- [x] Existing tests cover this change
- [ ] Not applicable

## Coverage notes

Wire fixtures are pinned verbatim from real Kimi Code 0.34.0 `wire.jsonl` logs (turn edges, `usage.record` with `usageScope:"turn"`, `llm.request` with provider-resolved model + alias), so upstream schema drift fails loudly in CI. Native/Python-only (no `web/` changes), so the e2e-ui judge's maintainer skip label applies.

## Changelog

Kimi sessions now report their model, token usage, and server-priced cost (at kimi's posted API rates when uncatalogued), deliver reliably into the web transcript, and can no longer strand a session `running` after a crash or interrupt.

## Post-consolidation update (2026-08-29)

- **Rebased onto current `main`** (5d08cb58a, ~276 commits absorbed); `kimi_auth.py` and tests now byte-identical to upstream's pay-per-use auth work.
- **Adversarial re-review round (2 independent reviewers, executed probes) found and fixed 4 defects** (this branch), each with a regression test failing on pre-fix code:
  - a dead pane no longer posts its failure fallback ahead of a stable newline-less `turn.ended(completed)` tail (no more false `failed → idle`);
  - generation resets clear `last_assistant_text` and sibling caches (a new turn can no longer surface the previous generation's answer);
  - `note_wire_restarted()` clears the pending-model attribution FIFO (usage can't bill under a dead generation's model);
  - the approval-hook budget is enforced on wall clock — connect time is reserved inside the read budget, keeping worst-case total at 598s under Kimi's 600s kill ceiling.
- Independent fix verification reproduced all four pre-fix symptoms and confirmed the fixes; 196 forwarder/hook tests green.

## 🎥 Demo

- [Demo video (44s, MP4)](https://github.com/btli/omnigent/releases/download/pr-demos-2026-08-29/demo-4439-kimi.mp4) — real kimi wire fixtures: `main` leaves a failed turn RUNNING forever with zero accounting; this branch surfaces the 401 as a failed edge, mirrors the model, posts token/cost usage ($0.2928 at posted K3 rates), and closes a dead pane's newline-less completed tail as idle.

## Issues

Closes #4437
Resolves [OMNI-2796](https://linear.app/omnigent/issue/OMNI-2796/kimi-native-sessions-silently-lose-dispatched-messages-strand-mid-turn)

## Issues

Resolves [OMNI-2822](https://linear.app/omnigent/issue/OMNI-2822/bug-kimikimi-native-sessions-never-report-model-token-usage-or-cost)

### CODE REVIEW

File: omnigent/model_fallbacks.py [hunk 1]
```diff
@@ -45,6 +45,22 @@ def static_model_fallback(provider_kind: str, cli: str) -> StaticModelFallback |
     return _STATIC_MODEL_FALLBACKS.get((provider_kind, cli))
 
 
+#: Kimi model families whose API list price is pinned in
+#: ``omnigent.llms.context_window._KIMI_POSTED_PRICING`` (matched by substring
+#: against the effective model id the kimi-native forwarder reports, e.g.
+#: ``system.ai.kimi-k3`` / ``kimi-k3-databricks``). Order matters: it keys the
+#: parallel rate table.
+KIMI_PRICED_MODEL_FAMILIES = StaticModelFallback(
```
Comment (TomeHirata) [inline omnigent/model_fallbacks.py]: Can we avoid hardcoded model ids?

Comment (btli) [reply omnigent/model_fallbacks.py]: Dropped the pinned table. The MLflow catalog already carries these models (moonshot.json: kimi-k3 / kimi-k2.7-code / kimi-k2.6; databricks.json: databricks-kimi-k3) at the same rates, so the real gap was id routing: the forwarder reports Databricks Unity Catalog ids like `system.ai.kimi-k3`, which the catalog lookup didn't recognise.

a0e29af78 fixes that at the lookup instead: `kimi-*` ids route to the moonshot catalog file, and `system.ai.<model>` maps to its `databricks-<model>` serving row. No model ids or prices are hard-coded; the same routing also resolves the context window for these ids. Covered by fixture-catalog routing tests (mutation-checked) and verified against the live catalog.

**Comment by btli** [comment]
If kimi stability is on the review table, these draft PRs are the rest of the picture. Each makes the kimi harness materially more usable; all are stuck in draft because fork PRs can no longer be flipped to ready via the API, so a maintainer click is needed on each. All four merge cleanly onto main today.

- #6449 — preseed workspace trust so the first native launch isn't blocked on Kimi 0.41's unhookable trust modal (also unblocks project MCP servers).
- #6370 — resolve kimi models from the Kimi CLI catalog; fixes a KeyError that reported "no usable model provider" and stops sub-agents inheriting the parent's Claude model.
- #6436 — route kimi permission prompts through the vendor-agnostic endpoint so approvals stop rendering as "Claude wants to call Bash", and fix a deny-keystroke race that could approve a denied tool.
- #6439 — Ctrl-S after Enter so a web message sent mid-generation steers the running turn instead of queueing.

**Review by TomeHirata — DISMISSED** [review]
LGTM

**Review by TomeHirata — APPROVED** [review]
